include/cantera/kinetics/Falloff.h

```diff
@@ -87,13 +126,143 @@ class Falloff
         return 0;
     }
 
+    virtual void setParameters(const AnyMap& node, const UnitStack& rate_units);
+
     //! Get the values of the parameters for this object. *params* must be an
     //! array of at least nParameters() elements.
+    /**
+     * @todo  deprecate; superseded by getFalloffCoeffs
+     */
     virtual void getParameters(double* params) const {}
 
-    //! Store the falloff-related parameters needed to reconstruct an identical
-    //! Reaction using the newReaction(AnyMap&, Kinetics&) function.
-    virtual void getParameters(AnyMap& reactionNode) const {}
+    virtual void getParameters(AnyMap& node) const;
+
+    //! Update information specific to reaction
+    //! @param shared_data  data shared by all reactions of a given type
+    virtual void updateFromStruct(const FalloffData& shared_data) {
+        updateTemp(shared_data.temperature, m_work.data());
+        m_rc_low = m_lowRate.evalFromStruct(shared_data);
+        m_rc_high = m_highRate.evalFromStruct(shared_data);
+        if (shared_data.finalized && m_rate_index != npos) {
+            m_thirdBodyConcentration = shared_data.conc_3b[m_rate_index];
+        }
+    }
+
+    //! Evaluate reaction rate
+    //! @param shared_data  data shared by all reactions of a given type
+    virtual double evalFromStruct(const FalloffData& shared_data) const {
+        double pr = m_thirdBodyConcentration * m_rc_low / (m_rc_high + SmallNumber);
+
+        // Apply falloff function
+        if (m_chemicallyActivated) {
+            // 1 / (1 + Pr) * F
+            pr = F(pr, m_work.data()) / (1.0 + pr);
+            return pr * m_rc_low;
+        }
+
+        // Pr / (1 + Pr) * F
+        pr *= F(pr, m_work.data()) / (1.0 + pr);
+        return pr * m_rc_high;
+    }
+
+    void check(const std::string& equation, const AnyMap& node);
+
+    //! Get buffered value of third-body concentration
+    const bool allowNegativePreExponentialFactor() const {
+        return m_negativeA_ok;
+    }
+
+    //! Get buffered value of third-body concentration
+    void setAllowNegativePreExponentialFactor(bool value) {
+        m_negativeA_ok = value;
+    }
+
+    //! Get buffered value of third-body concentration
+    const double thirdBodyConcentration() const {
```

Comment:
I don't think it's necessary (and sometimes can be unhelpful) to declare the return type as `const` unless returning a pointer or reference. This applies to a few other methods here, like `allowNegativePreExponentialFactor` and `chemicallyActivated`.